Add feature to extract CSV files from LEIE site

kevinwan1996 opened the pull request:
Automates downloading the UPDATED, EXCLUDED, and REINSTATED .csv files from the LEIE website. Can choose where to save these files and allow the ETL code to get information from this location.

Tested it locally just by setting a path and checking to see whether the .csv files saved were in the correct location and not altered.

brainwane commented on etl/leie/leie/download_csv.py:
```diff
@@ -0,0 +1,50 @@
+from bs4 import BeautifulSoup
```
Please add a `#!` line here to indicate that this is a Python file (and, if necessary, what version of Python it uses), and please add a comment explaining what the script does. You can look at https://github.com/OpenTechStrategies/psm/blob/master/etl/leie/leie/etl.py for an example.

brainwane commented on etl/leie/leie/download_csv.py:
```diff
@@ -0,0 +1,50 @@
+	for ul in uls:
+		lis = ul.findAll('li')
+		for li in lis:
+			if 'UPDATED.csv' in li.find('a')['href'] or 'EXCL.csv' in li.find('a')['href'] or 'REIN.csv' in li.find('a')['href']:
```
It would be great for you to add a comment here explaining why it's looking for these particular substrings, and why we can't just `wget` the same filenames over and over.

brainwane commented on etl/leie/leie/download_csv.py:
```diff
@@ -0,0 +1,50 @@
+import urllib
+
+#Enter in the paths to save the files to. Enter in order: Updated, Excluded, Reinstated
+def download_csv(list_of_paths):
```
As far as I can tell, you know that we will be downloading three files every time we run this script. Can you help me understand why this function should take a list of URLs (`list_of_paths`), instead of simply running with no arguments?

brainwane commented on etl/leie/leie/download_csv.py:
```diff
@@ -0,0 +1,50 @@
+
+
+if __name__=='__main__':
+	urls = ['test1.csv','test2.csv','test3.csv']
```
Please see my comment above about whether this is a necessary intermediate step. Also, it would be great to store the files with more descriptive filenames, like "updated-as-of-YEAR-MONTH-DAY", "exclusion-supplement-from-YEAR-MONTH", and "reinstatements-from-YEAR-MONTH".

brainwane commented on etl/leie/leie/download_csv.py:
```diff
@@ -0,0 +1,50 @@
+if __name__=='__main__':
+	urls = ['test1.csv','test2.csv','test3.csv']
+	download_csv(urls)
+	# save_file()
```
You can remove this -- I presume you used to run `save_file` here but no longer do.

brainwane commented:
@kevinwan1996 Also, when update this pull request, you don't need to close the old one and open a new one. You can [update your pull request](https://zulip.readthedocs.io/en/latest/git-guide.html#update-a-pull-request) and we recommend you do that (instead of closing the old PR and opening a new one) because then it's easier to see the history of what you've done, to to check that you've responded to previous reviewers' comments, and to preserve links we've mentioned (in comments) with other issues and PRs.

brainwane commented:
One more comment, Kevin: could you `chmod` the `download_csv.py` file to make it executable? `chmod u+x download_csv.py` should do it. Thanks!

kevinwan1996 commented:
Thank you for the feedback, I really appreciate it! I just changed the code based on your comments. 

I originally had list_of_paths because I wanted to allow us to set the destination for where to save the csv files that were downloaded. I probably should've named it list_of_destinations, but I changed it to just a 'directory' parameter. I thought that all three files would be added to the same directory, so you can just input a directory in and it will save the file there. Please let me know if I should change this.

The reason why I had the substrings and didn't just use a url request was because the url changes monthly. Since the script only needs to be run once a month, it wouldn't be too much work just to change the url, but I thought it would be easier to automatically find the url so we wouldn't have to change the script every month. 

The naming convention for the saved files are now 06-2017-Updated-LEIE-Database.csv, 06-2017-Exclusions.csv, and 06-2017-Reinstatements.csv). I just pulled the text straight from the site to name the files. Please let me know if there is a better naming convention you have in mind!

I also chmod'd the file, so that should be executable now!

brainwane commented on etl/leie/leie/download_csv.py:
```diff
@@ -15,36 +20,39 @@ def download_csv(list_of_paths):
-	urls = ['test1.csv','test2.csv','test3.csv']
-	download_csv(urls)
-	# save_file()
+# if __name__=='__main__':
```
You probably didn't mean to comment out the `if __name__=='__main__'` line. :)

kevinwan1996 commented:
Super sorry I thought this was complete already! Please let me know if there is anything else I should change with this script.

jvasile commented:
ETL.py already downloads UPDATED.csv.  The patches I submitted today in etl-refactor and etl-download-reinstatements add the ability to download the other files.  In the context of the ETL stuff, we should use that new code-- it's more robust and it connects to the process for loading that data into the database and serve it over an API.

This script has the benefit of being simple and not connecting to anything.  If it belongs in the tree, maybe it belongs in a scripts directory so people can use it if they want.  Best to make it clear that it's not part of the toolchain though.

brainwane commented:
@kevinwan1996 I agree with @jvasile that this ought to live elsewhere -- I think the `scripts/` directory (in the root of the repository) would work well. I'll take another look now at the rest of the script to see if there's anything else to fix.

brainwane commented on etl/leie/leie/download_csv.py:
```diff
@@ -0,0 +1,58 @@
+
+	#find the hrefs that contain the 3 different csv files
+	'''these tags hold the download links, which change every month. It's easier just to find 
+	automatically find them each time rather than alter the script each month'''
```
You have a repeated word here. Here's an improved version of the comment:

These tags hold the download links. The filenames change every month (e.g., https://oig.hhs.gov/exclusions/downloadables/2017/1706EXCL.csv) so we automatically find them by looking for the substring that, as far as we can tell, doesn't change.

brainwane commented on etl/leie/leie/download_csv.py:
```diff
@@ -0,0 +1,58 @@
+	uls = content.findAll('ul')
+
+	csv_urls = []
+	csv_names = []
```
Since we process each `csv_url` element in tandem with its matching `csv_names` element, maybe you should use a dictionary instead, to hold and more easily manipulate the key-value pairs. This is an optional thing for you to change, just to improve the quality of the code.

kfogel replied:
Yes, either a dictionary or a nested list (e.g., a list whose elements are two-element lists, each of which represents a matching pair).

brainwane commented on etl/leie/leie/download_csv.py:
```diff
@@ -0,0 +1,58 @@
+			csv_names[i] = directory + '-'.join(csv_names[i].split(' ')) + '.csv'
+			save_file(csv_urls[i], csv_names[i])
+	else:
+		raise ValueError("Number of CSVs found was incorrect")
```
`download_csv` is pretty long and does several things. I suggest you break it into about three functions (to aid in future debugging), maybe fewer, maybe more; use your judgment.

OIG might completely change their file-naming customs, or switch what page they're hosting their CSVs on. Could you add a function, or a conditional within one of the functions, to check for a situation where none of the `href`s on the page point to a `.csv` file, and return an error saying that that's the problem?

jvasile replied:
I like the comment about future-proofing especially.

kfogel commented on etl/leie/leie/download_csv.py:
```diff
@@ -0,0 +1,58 @@
+
+
+if __name__=='__main__':
+	download_csv('/Users/Kevin/Documents/')
```
Typical behavior, I think, would be either the current working directory, or a destination directory specified on the command line.  Unless there is a convention for a particular data layout under `etl/`, in which case then the solution is to pass the path to the `etl/` directory on the command line.

kevinwan1996 commented:
Hello! I made the changes based on the comments, but I had some trouble moving the script to the scripts directory cause my fork didn't have that folder yet. I tried following along this https://stackoverflow.com/questions/7244321/how-do-i-update-a-github-forked-repository to update my fork, but I couldn't get it to work so I thought I could just delete and re-fork my fork. Unfortunately I think this made it so that the pull request doesn't automatically update with my commits. I'm really sorry for the inconvenience and is there anything I could do to fix this? I've just attached the new file to this comment as a txt file
[download_csv.txt](https://github.com/OpenTechStrategies/psm/files/1156341/download_csv.txt)
.

brainwane commented:
@kevinwan1996 it's common to run into difficulty when you're starting to use Git and you don't understand what it's doing under the hood. I recommend you read [this explanation of the Git model](http://web.mit.edu/nelhage/Public/git-slides-2009.pdf) and do some practice on a toy repository to get the hang of things better.

In the meantime, I'll set up a new PR with the file that you've attached and take care of merging it. Thanks.

kevinwan1996 commented:
Thank you for the link and all the help! I'll be sure to read it asap

brainwane commented:
Closing this PR in favor of #277. Thanks @kevinwan1996.
